[1.21] Add holder helpers to LevelReader and HolderLookupProvider

baileyholl opened the pull request:
Numerous mods have found themselves in wrapper hell, and it seems like everyone is making utility methods such as 
```java
public class HolderHelper {

    public static <T> Holder<T> unwrap(Level level, ResourceKey<T> key){
        return level.registryAccess().registryOrThrow(key.registryKey()).getHolderOrThrow(key);
    }
}
```
to cut down on the verbosity. Bringing it closer to level instead lets us easily unwrap objects we expect to be present, such as enchantments or our own keys.
```java
level.holder(Enchantments.SILK_TOUCH)
```

baileyholl committed:
Add holder helpers to ILevelExtension

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=1198) <br/>All committers have signed the CLA.

shartte commented:
Wouldn't the highest possible level for this be a `ILevelReaderExtension`?

baileyholl commented:
> Wouldn't the highest possible level for this be a `ILevelReaderExtension`?

Ah that is a good idea. Is it worth making a new extension file exclusively for this? I am open to it if we aren't being stingy with the interface extensions 😄

shartte commented:
I'd also consider not adding this to level, but maybe rather `net.minecraft.core.HolderLookup.Provider`, since that is one method away and is passed around mercilessly around the code-base now (i.e. Tooltips, NBT serialization/deserialization).

I.e., having these methods on Level doesn't help you in those cases, when you for example want to get an Enchantment.

baileyholl committed:
Add to HolderLookupProvider, move ILevelExtensions to LevelReader

baileyholl commented:
> I'd also consider not adding this to level, but maybe rather `net.minecraft.core.HolderLookup.Provider`, since that is one method away and is passed around mercilessly around the code-base now (i.e. Tooltips, NBT serialization/deserialization).
> 
> I.e., having these methods on Level doesn't help you in those cases, when you for example want to get an Enchantment.

Good idea. 
I am in favor of having both, as I would love to save the extra `level.registryAccess` step. 

Adding it to Provider allows us to now fetch the holder from `TooltipContext` via
```java
 var enchant = tooltipContext.registries().holderOrThrow(Enchantments.FORTUNE)
```

I moved the extensions from ILevel to ILevelReader, and added the same methods to a IHolderLookupProvider extension.

baileyholl committed:
generate patches

baileyholl committed:
Move existing hasChunkAt patch to interface

baileyholl commented on src/main/java/net/neoforged/neoforge/common/extensions/ILevelReaderExtension.java:
```diff
@@ -0,0 +1,40 @@
+        return (LevelReader) this;
+    }
+
+    default boolean isAreaLoaded(BlockPos center, int range) {
```
Moved over from the existing patch, as this can now live entirely in the interface.

baileyholl committed:
Implement RegistryAccess via delegation to expose holder methods on level

Matyrobbrt commented on patches/net/minecraft/core/HolderLookup.java.patch:
```diff
@@ -1,5 +1,14 @@
+     }
+ 
+-    public interface Provider {
++    public interface Provider extends net.neoforged.neoforge.common.extensions.IHolderLookupProviderExtension{
```
Please add a new line between the interface and the left brace.

baileyholl replied:
A new line or a space? The existing file had a space, but not a new line. Other interfaces do not add a new line for the brace like in IlevelExtension
```java
public abstract class Level extends net.neoforged.neoforge.attachment.AttachmentHolder implements LevelAccessor, AutoCloseable, net.neoforged.neoforge.common.extensions.ILevelExtension {
```

Matyrobbrt replied:
Extra space yes, sorry.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/extensions/IHolderLookupProviderExtension.java:
```diff
@@ -0,0 +1,26 @@
+import net.minecraft.resources.ResourceKey;
+
+public interface IHolderLookupProviderExtension {
+    default HolderLookup.Provider self() {
```
Should be private.

Matyrobbrt commented on src/main/java/net/neoforged/neoforge/common/extensions/ILevelReaderExtension.java:
```diff
@@ -0,0 +1,39 @@
+import net.minecraft.world.level.LevelReader;
+
+public interface ILevelReaderExtension extends RegistryAccess {
+    default LevelReader self() {
```
Should be private.

Matyrobbrt reviewed (changes requested):
I'd add a bit of basic documentation on the extension methods.

baileyholl committed:
Docs, patch formatting

Technici4n commented on src/main/java/net/neoforged/neoforge/common/extensions/ILevelReaderExtension.java:
```diff
@@ -0,0 +1,40 @@
+     * directly without traversing through level.registryAccess().
+     * For additional utilities, see {@link IHolderLookupProviderExtension}
+     */
+
```
Why is there an empty line here?

baileyholl replied:
Mainly to keep it away from the javadoc directly on the method, as its more of a note for maintainers instead of modders.

Technici4n replied:
It's still part of the method javadoc if it's a javadoc comment, even with additional empty lines.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/extensions/ILevelReaderExtension.java:
```diff
@@ -0,0 +1,40 @@
+     */
+
+    @Override
+    default <E> Optional<Registry<E>> registry(ResourceKey<? extends Registry<? extends E>> p_123085_) {
```
I'd give the parameter a proper name, for example `registryKey`.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/extensions/ILevelReaderExtension.java:
```diff
@@ -0,0 +1,40 @@
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.LevelReader;
+
+public interface ILevelReaderExtension extends RegistryAccess {
```
Not a fan of making it extend `RegistryAccess` directly, as that brings weird methods like `freeze()` in scope.

shartte replied:
Uh yes that can have unforeseen consequences. 
Before we do that I'd rather have a copy of the two util methods in here

baileyholl replied:
Previously I had the methods duplicated. I believe it could be narrowed in scope to only extend HolderLookup.Provider, which afaik is sufficient for the majority of usecases.

Technici4n replied:
A bit of duplication is fine. (And generally preferable to inheritance spaghetti, IMO).

baileyholl replied:
While im not sure how useful or relevant it might be, it would also let you pass `level` in place of methods that require a Provider. Since levels are typically the only way to obtain a registry in many scopes to begin with, I don't think its *too* smelly to allow a level to be treated as a provider directly, but I won't die on the hill, and those use cases are far fewer anyway. If duplicated methods are better for maintainability im happy with that.

Technici4n replied:
I don't think there is much value when one could instead just pass `level.registries()`.

baileyholl replied:
Sounds good. I removed the extension and added the methods to `IlevelReaderExtension` which calls those same methods from the HolderExtension.

baileyholl committed:
Remove level holder extension, add duplicate methods and docs
